File: jscomp/test/gpr_3697_test.js
'use strict';

var CamlinternalLazy = require("../../lib/js/camlinternalLazy.js");

function fix(param) {
  return /* Fix */{
          _0: {
            RE_LAZY_DONE: false,
            value: (function () {
                return fix(undefined);
              })
          }
        };
}

function unfixLeak(_f) {
  while(true) {
    var f = _f;
    _f = CamlinternalLazy.force(f._0);
    continue ;
  };
}

function unfix(p) {
  while(true) {
    var match = p.contents;
    p.contents = CamlinternalLazy.force(match._0);
  };
  
}

exports.fix = fix;
exports.unfixLeak = unfixLeak;
exports.unfix = unfix;
/* No side effect */

File: jscomp/test/gpr_3697_test.ml
type 'a t = Fix of 'a t lazy_t

let rec fix () = Fix (lazy (fix ()))


let rec unfixLeak (Fix f) =
  unfixLeak @@ Lazy.force f


let  unfix p = 
  while true do 
    p := match !p with (Fix lazy h) -> h
  done 
(* ;; unfixLeak (fix ())   *)

(* ;; unfix (ref (fix ())) *)